libmamba/src/core/problems_graph.cpp

```diff
@@ -0,0 +1,493 @@
+    {
+        m_deps.insert(edge.m_dep);
+    }
+
+    bool MGroupEdgeInfo::operator==(const MGroupEdgeInfo& edge)
+    {
+        return m_deps == edge.m_deps;
+    }
+
+    /************************
+     * problems graph implementation *
+     *************************/
+
+
+    MProblemsGraphs::MProblemsGraphs(MPool* pool)
+        : m_pool(pool)
+    {
+    }
+
+    MProblemsGraphs::MProblemsGraphs()
+    {
+    }
+
+    MProblemsGraphs::MProblemsGraphs(MPool* pool, const std::vector<MSolverProblem>& problems)
+    {
+        create_graph(problems);
+    }
+
+    MProblemsGraphs::merged_conflict_graph MProblemsGraphs::create_graph(
+        const std::vector<MSolverProblem>& problems)
+    {
+        create_initial_graph(problems);
+        create_unions();
+        return create_merged_graph();
+    }
+
+    void MProblemsGraphs::create_initial_graph(const std::vector<MSolverProblem>& problems)
+    {
+        for (const auto& problem : problems)
+        {
+            std::cerr << problem.to_string() << std::endl;
+        }
+        std::cerr << "----" << std::endl;
+        for (const auto& problem : problems)
+        {
+            add_problem_to_graph(problem);
+        }
+    }
+
+    void MProblemsGraphs::add_problem_to_graph(const MSolverProblem& problem)
+    {
+        std::cerr << problem.to_string() << std::endl;
+        std::cerr << problem.type << " " << get_value_or(problem.source()) << " "
+                  << get_value_or(problem.target()) << " " << problem.dep().value_or("None")
+                  << std::endl;
+        switch (problem.type)
+        {
+            case SOLVER_RULE_PKG_CONSTRAINS:  // source_id -> dep_id -> <conflict>
+            {
+                if (!has_values(problem, { problem.source(), problem.target() })
+                    || !has_values(problem, { problem.dep() }))
+                {
+                    break;
+                }
+                auto source = MNode(problem.source().value());
+                std::vector<Id> constrained_ids = m_pool->select_solvables(problem.dep_id);
+                for (const auto& constraint_id : constrained_ids)
+                {
+                    PackageInfo constraint_pkg
+                        = PackageInfo(pool_id2solvable(*m_pool, constraint_id));
+                    // std::cerr << "target = " << constraint_pkg.str() << " " << problem.dep() <<
+                    // std::endl;
+                    add_conflict_edge(source, MNode(constraint_pkg), problem.dep().value());
+                }
+                node_id source_id = get_or_create_node(source);
+                node_id target_id = get_or_create_node(MNode(problem.target().value()));
+                add_solvables_to_conflicts(source_id, target_id);
+                break;
+            }
+            case SOLVER_RULE_PKG_REQUIRES:  // source_id -> dep_id -> target_id
+            case SOLVER_RULE_JOB:           // entry point for our solver
```

Comment:
Yes! I discard the source_id & target_id in the case of SOLVER_RULE_JOB. I only use the dep_id because I want to get the edge info with the deps the user requested. 
Sometimes, though, the dep_id of SOLVER_RULE_JOB is invalid and it seg faults when we are querying the pool - I saw that when we have the rule SOLVER_RULE_UPDATE. in the case of mamba install jupyterlab>3.0 openssl=3.*
I think we need to discuss with libsolv and either get something correct from their side (ideally) or clean it up on our side (ignore dep_ids the user did not request).